Stargate blog (#164)
ippontech/blog-usa: posts/stargate.md

---
authors:
- Theo Lebrun
tags:
- Cassandra
- Big Data
date: 2021-01-14T14:50:55.000Z
title: "Store and query any data effortlessly using Stargate"
image:
---

> [Stargate](https://stargate.io/) is one of the latest shiny tools from [DataStax](https://www.datastax.com/) that will act as a data gateway to abtract Cassandra-specific concepts and remove barriers of entry for new software developers!

Reviewer: ```suggestion
[Stargate](https://stargate.io/) is one of the latest shiny tools from [DataStax](https://www.datastax.com/) that will act as a data gateway to abstract Cassandra-specific concepts and remove barriers of entry for new software developers!
```

> Stargate is a full open source project ([GitHub repository](https://github.com/stargate/stargate)) that supports the most popular way of querying data (REST, GraphQL, Document and CQL API). By chosing their favorite way of querying data, developers will spend less time learning the API and more time on implementing new features.

Reviewer: ```suggestion
Stargate is a full open source project ([GitHub repository](https://github.com/stargate/stargate)) that supports the most popular ways of querying data (REST, GraphQL, Document and CQL API). By choosing their favorite way of querying data, developers will spend less time learning the API and more time on implementing new features.
```

# Docker and Stargate for an easy setup

Thanks to [Docker](https://www.docker.com/), setting up Stargate is very simple and will not even require a Cassandra cluster. Here is the command that will start the Stargate container:

```bash
docker run --name stargate \
  -p 8080:8080 \
  -p 8081:8081 \
  -p 8082:8082 \
  -p 127.0.0.1:9042:9042 \
  -d \
  -e CLUSTER_NAME=stargate \
  -e CLUSTER_VERSION=3.11 \
  -e DEVELOPER_MODE=true \
  stargateio/stargate-3_11:v1.0.0
```

The flag `DEVELOPER_MODE` tells Stargate to use its own Cassandra cluster but you can also use your own cluster. In order to provide your own cluster, you can follow the [official documentation](https://stargate.io/docs/stargate/1.0/developers-guide/install/install_existing_cstar.html).

The Stargate container exposes 4 different services:
- GraphQL interface for CRUD (**8080**)
- REST authorization service for generating tokens (**8081**)
- REST interface for CRUD (**8082**)
- CQL service (**9042**)

> Make sure that the container was started without any errors before continuing this blog.

Reviewer: ```suggestion
Make sure that the container was started without any errors before continuing.
```

![Output](https://raw.githubusercontent.com/Falydoor/blog-usa/stargate/images/2021/01/stargate-1.png)

Here is the table definition of `vehicle`:

```json
{
  "data": {
    "name": "vehicle",
    "keyspace": "blog",
    "columnDefinitions": [
      {
        "name": "manufacturer",
        "typeDefinition": "varchar",
        "static": false
      },
      {
        "name": "type",
        "typeDefinition": "varchar",
        "static": false
      },
      {
        "name": "color",
        "typeDefinition": "varchar",
        "static": false
      },
      {
        "name": "model",
        "typeDefinition": "varchar",
        "static": false
      },
      {
        "name": "vin",
        "typeDefinition": "varchar",
        "static": false
      }
    ],
    "primaryKey": {
      "partitionKey": [
        "manufacturer"
      ],
      "clusteringKey": [
        "type"
      ]
    },
    "tableOptions": {
      "defaultTimeToLive": 0,
      "clusteringExpression": [
        {
          "order": "ASC",
          "column": "type"
        }
      ]
    }
  }
}
```

> The partition key is `manufacturer` and the clustering key is `type` so we can get all vehicles by manufacturer and sort them by type.

Reviewer: Is that a "clustering key" or "sort key"?

Author: Clustering key yeah

# Querying the data

## GraphQL API

> Stargate will provide a GraphQL playground available at http://localhost:8080/playground, don't forget to set the header `x-cassandra-token` with the correct value at the bottom of the UI.

Reviewer: ```suggestion
Stargate will provide a GraphQL playground available at http://localhost:8080/playground and don't forget to set the header `x-cassandra-token` with the correct value at the bottom of the UI.
```

Data can be queried by opening a new tab and then using `http://localhost:8080/graphql/blog` for the schema.

Here is a query that will retrieve all vehicles manufactured by `Tesla`:

```
query vehicles {
  vehicle(value: { manufacturer: "Tesla" }) {
    values {
      manufacturer
      type
      color
      model
      vin
    }
  }
}
```

The query should output something like below:

![Output](https://raw.githubusercontent.com/Falydoor/blog-usa/stargate/images/2021/01/stargate-2.png)

> What makes GraphQL great is that you can define which fields to retrieve directly in your query to simplifies things a lot when querying complex APIs.

Reviewer: ```suggestion
What makes GraphQL great is that you can define which fields to retrieve directly in your query to simplify things a lot when querying complex APIs.
```

More details on how to use the GraphQL API can be found [here](https://stargate.io/docs/stargate/1.0/developers-guide/graphql-using.html).

## REST API

> The query below can also be done using a more traditional way with the REST API and using predefined search terms.

Reviewer: ```suggestion
The query below can also be done in a more traditional way with the REST API and using predefined search terms.
```

```bash
curl --location \
--request POST 'localhost:8082/v2/namespaces/blog/collections/vehicles' \
--header "X-Cassandra-Token: $AUTH_TOKEN" \
--header 'Content-Type: application/json' \
--data '{
  "vin": "7OS5T5HQLGT073117",
  "model": "Spyder",
  "type": "Sedan",
  "color": "teal",
  "manufacturer":"Tesla"
}'
```

And then retrieve it:

```bash
curl --location \
--request GET 'localhost:8082/v2/namespaces/blog/collections/vehicles' \
--header "X-Cassandra-Token: $AUTH_TOKEN" \
--header 'Content-Type: application/json'
```

But what makes Stargate awesome is that I can also use GraphQL to retrieve my document:

```
query vehicles {
  vehicles(value: {key: "DOCUMENT_ID"}) {
    values{
      leaf
      text_value
    }
  }
}
```

The query output might be confusing because the document is flattened but a condition can be added to only retrieve the needed field.

More details on how to use the Document API can be found [here](https://stargate.io/docs/stargate/1.0/developers-guide/document-using.html).

## CQL API

Finally, the CQL API will let you access the data using your favorite native language drivers like you would normally do with a regular Cassandra cluster.

> In our case, after being connected to our Stargate container following the [instructions here](https://stargate.io/docs/stargate/1.0/developers-guide/cql-using.html), we can run the following query to get all our vehicles:

Reviewer: ```suggestion
In our case, after being connected to our Stargate container using the [instructions here](https://stargate.io/docs/stargate/1.0/developers-guide/cql-using.html), we can run the following query to get all our vehicles:
```
